### PR TITLE
Update of L1 tracking (matching PR #30168)

### PR DESCRIPTION
**Pull request by skinnari**
This is an identical update as PR #30168 (for reference, description below) but to CMSSW_11_2_X instead of CMSSW_11_1_X . 

------
This PR contains fixes for the L1 tracking for the HLT TDR:

(1) The update restores the expected pt resolution for high-pt particles (a bug had been introduced during the CMSSW integration).
(2) It allows to run the displaced ("extended") version of the L1 tracking w/o having to be in a particular directory (using edm::fileInPath), to make sure it could run in central production.
------

### CODE REVIEW

**Comment by fwyzard** [comment]
type bugfix

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by silviodonato** [comment]
assign upgrade

**Comment by silviodonato** [comment]
urgent

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc [hunk 1]
```diff
@@ -453,20 +455,16 @@ void TrackletProcessor::execute() {
...
-                                   (iphiinnerbin.value() << outerphibits_) + iphiouterbin.value();
+              unsigned int index = (iphiinnerbin.value() << outerphibits_) + iphiouterbin.value();
 
-              assert(index < phitable_[phiindex].size());
-              if (!phitable_[phiindex][index]) {
-                if (settings_.debugTracklet()) {
-                  edm::LogVerbatim("Tracklet") << "Stub pair rejected because of tracklet pt cut";
-                }
-                continue;
+              if (iSeed_ >= 4) {  //Also use r-position
```
Comment (kpedro88) [inline L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc:465]: magic numbers should be named constants (4, 3, 1, 2, another 3...)

File: L1Trigger/TrackFindingTracklet/src/TripletEngine.cc [hunk 2]
```diff
@@ -243,7 +243,7 @@ void TripletEngine::execute() {
       }
 
       else if (disk2_ == 2 && layer3_ == 2) {
-        int lookupbits = (int)((firstvmstub.vmbits().value() >> 9) & 1023);
+        int lookupbits = (int)((firstvmstub.vmbits().value() >> 10) & 1023);
```
Comment (kpedro88) [inline L1Trigger/TrackFindingTracklet/src/TripletEngine.cc:246]: still lots of magic numbers here

File: L1Trigger/TrackFindingTracklet/src/TripletEngine.cc [hunk 3]
```diff
@@ -288,12 +289,12 @@ void TripletEngine::execute() {
...
                 if (settings_.debugTracklet()) {
                   edm::LogVerbatim("Tracklet")
-                      << "Stub pair rejected because of stub pt cut bends : "
+                      << "Stub triplet rejected because of stub pt cut bends : "
                       << benddecode(secondvmstub.bend().value(), secondvmstub.isPSmodule()) << " "
                       << benddecode(thirdvmstub.bend().value(), thirdvmstub.isPSmodule());
                 }
-                if (!settings_.writeTripletTables())
-                  continue;
+                //if (!settings_.writeTripletTables())
```
Comment (kpedro88) [inline L1Trigger/TrackFindingTracklet/src/TripletEngine.cc:296]: delete commented-out code

File: L1Trigger/TrackFindingTracklet/src/TripletEngine.cc [hunk 4]
```diff
@@ -377,8 +377,8 @@ void TripletEngine::execute() {
...
```
Comment (kpedro88) [inline L1Trigger/TrackFindingTracklet/src/TripletEngine.cc:380]: delete commented-out code

File: L1Trigger/TrackFindingTracklet/src/VMRouterTable.cc [hunk 5]
```diff
@@ -211,8 +211,9 @@ int VMRouterTable::getLookup(unsigned int layerdisk, double z, double r, int ise
     int rbin2 = NBINS * (rmax - settings_.rmindiskvm()) / (settings_.rmaxdiskvm() - settings_.rmindiskvm());
 
     if (iseed == 10) {
-      rbin1 = NBINS * (rmin - settings_.rmindiskvm()) / (settings_.rmaxdisk() - settings_.rmindiskvm());
-      rbin2 = NBINS * (rmax - settings_.rmindiskvm()) / (settings_.rmaxdisk() - settings_.rmindiskvm());
+      double rminspec = 40.0;
```
Comment (kpedro88) [inline L1Trigger/TrackFindingTracklet/src/VMRouterTable.cc:214]: constexpr

File: L1Trigger/TrackFindingTracklet/test/L1TrackNtupleMaker_cfg.py [hunk 6]
```diff
@@ -111,22 +113,24 @@ def getTxtFile(txtFileName):
...
     L1TRK_NAME  = "TTTracksFromTrackletEmulation"
     L1TRK_LABEL = "Level1TTTracks"
+    L1TRUTH_NAME = "TTTrackAssociatorFromPixelDigis"
 
 # LEGACY ALGORITHM (EXPERTS ONLY): TMTT  
 elif (L1TRKALGO == 'TMTT'):
     print "\n WARNING - this is not a recommended algorithm! Please use HYBRID (HYBRID_DISPLACED)! \n"
     process.load("L1Trigger.TrackFindingTMTT.TMTrackProducer_Ultimate_cff")
-    L1TRK_PROC  =  process.TMTrackProducer
+    L1TRK_PROC  =  process.TMTrackProducer
```
Comment (kpedro88) [inline L1Trigger/TrackFindingTracklet/test/L1TrackNtupleMaker_cfg.py:122]: remove unnecessary whitespace at end of line

**Comment by kpedro88** [comment]
@skinnari what is the status of this PR

**Comment by silviodonato** [comment]
@skinnari we are going to close CMSSW_11_1_0, it is very important to have this PR updated and merged

File: L1Trigger/TrackFindingTracklet/src/TripletEngine.cc (hunk 4 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/src/TripletEngine.cc:380]: done

File: L1Trigger/TrackFindingTracklet/src/TripletEngine.cc (hunk 3 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/src/TripletEngine.cc:296]: done

File: L1Trigger/TrackFindingTracklet/src/TripletEngine.cc (hunk 2 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/src/TripletEngine.cc:246]: made named

File: L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc (hunk 1 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/src/TrackletProcessor.cc:465]: made named

File: L1Trigger/TrackFindingTracklet/src/VMRouterTable.cc (hunk 5 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/src/VMRouterTable.cc:214]: done

File: L1Trigger/TrackFindingTracklet/test/L1TrackNtupleMaker_cfg.py (hunk 6 above)
Comment (skinnari) [reply L1Trigger/TrackFindingTracklet/test/L1TrackNtupleMaker_cfg.py:122]: done

**Comment by skinnari** [comment]
@silviodonato sorry, OK, updating this imminently!

**Comment by silviodonato** [comment]
please test
Thanks Louise, all comments have been addressed.
Please update also the backport to 11_1_X https://github.com/cms-sw/cmssw/pull/30168 (or create a new PR)

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
@kmcdermo @rekovic can we merge this PR? The standalone test is ok (see https://github.com/cms-sw/cmssw/pull/30245#issuecomment-650352515)

**Comment by kpedro88** [comment]
+upgrade

**Comment by silviodonato** [comment]
merge
cc @rekovic

**Comment by skinnari** [comment]
> -1
> @skinnari This pull request cannot be automatically merged, could you please rebase it?
> 
> You can see the log for git cms-merge-topic here:
> https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-30245/7502/git-merge-result

just to be sure - there was a message overnight that I needed to rebase for it to be able to be merged, but it seems here that it is listed as "merged". do I need to rebase or not?

**Comment by silviodonato** [comment]
@skinnari the message was related to another PR. We tested several PRs together, the test failed because there was a conflict between two of them, and then the message was sent to the author of each PR. In short, please ignore that message.
